File: csharp/triangle/Triangle.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Triangle
{
    private static int NumberOfEqualSides(IEnumerable<decimal> sides)
    {
        var equalSides = 4 - sides.Distinct().Count();
        return equalSides == 1 ? 0 : equalSides;
    }

    public static TriangleKind Kind(decimal side1, decimal side2, decimal side3) {

        var sides = new[] { side1, side2, side3 };

        if (sides.Any(x => x <= 0) || sides.Sum() - sides.Max() <= sides.Max())
        {
            throw new TriangleException();
        }

        return (TriangleKind) NumberOfEqualSides(sides);
    }
}

File: swift/triangle/Triangle.swift
struct Triangle {
    let sides : [Double]

    init(_ side1 : Double, _ side2 : Double, _ side3 : Double) {
        sides = [side1, side2, side3]
    }

    var kind : String {
        let sidesMax : Double! = sides.max()

        if sides.filter({ $0 <= 0 }).count == 0 && (sides.reduce(0, +) - sidesMax) > sidesMax {
            switch getNumberOfEqualSides() {
                case 3:
                    return "Equilateral"
                case 2:
                    return "Isosceles"
                case 0:
                    return "Scalene"
                default:
                    break
            }
        }

        return "ErrorKind"
    }

    func getNumberOfEqualSides() -> Int {
        var equalSides = 4 - Set(sides).count
        return equalSides == 1 ? 0 : equalSides
    }
}

